feat: dataspace protocol transfer-process-transformer

janpmeyer opened the pull request:
## What this PR changes/adds

Implements transfer-process transformers following :dsp:dsp-catalog.

`:data-protocols:dsp:dsp-transfer-process:dsp-transfer-process-transform`

## Why it does that

Support the dataspace protocol specification.

## Further notes


## Linked Issue(s)

Relates #2474

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

juliapampus commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/DspTransferProcessTransformExtension.java:
```diff
@@ -0,0 +1,60 @@
+    public void initialize(ServiceExtensionContext context) {
+        var builderFactory = Json.createBuilderFactory(Map.of());
+
+        var mapper = typeManager.getMapper("json-ld");
```
Should be `JsonLdExtension.TYPE_MANAGER_CONTEXT_JSON_LD`

juliapampus commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferCompletionMessageTransformer.java:
```diff
@@ -0,0 +1,56 @@
+    public JsonObjectFromTransferCompletionMessageTransformer(JsonBuilderFactory jsonBuilderFactory, ObjectMapper mapper) {
+        super(TransferCompletionMessage.class, JsonObject.class);
+        this.jsonBuilderFactory = jsonBuilderFactory;
+        this.mapper = mapper;
```
unused mapper

juliapampus replied:
see also for other tranformer classes

juliapampus commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,85 @@
+        var builder = jsonBuilderFactory.createObjectBuilder();
+
+        builder.add(JsonLdKeywords.ID, String.valueOf(UUID.randomUUID()));
+        builder.add(JsonLdKeywords.TYPE, DSPACE_SCHEMA + "TransferRequestMessage");
```
Could be a property in DspCatalogPropertyAndTypeNames (same for the other messages)

ndr-brt commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/test/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/JsonObjectFromTransferProcessTransformerTest.java:
```diff
@@ -0,0 +1,242 @@
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_TRANSFER_TERMINATION_TYPE;
+
+
+class JsonObjectFromTransferProcessTransformerTest {
```
I'm not too against this "integrated" way of testing transformers, but for the catalog they were tested in an "unitary" way, I would keep the same convention here as well

jimmarino replied:
I think each transformer needs a unit test. Otherwise, it is too difficult to maintain them. Please follow the existing conventions we have.

janpmeyer replied:
Test are extracted.

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/build.gradle.kts:
```diff
@@ -0,0 +1,38 @@
+
+    api(project(":data-protocols:dsp:dsp-http-spi"))
+
+    implementation("jakarta.json:jakarta.json-api:2.1.1")
+    implementation("org.eclipse.parsson:parsson:1.1.1")
```
Why are we introducing an external library here? What is its purpose and license, and why do we need it?

janpmeyer replied:
Dependencies are removed

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/DspCatalogPropertyAndTypeNames.java:
```diff
@@ -0,0 +1,53 @@
+    String DSPACE_STATE_TYPE = DSPACE_SCHEMA + "state";
+
+    String DCT_FORMAT = DCT_SCHEMA + "format";
+}
```
Aren't many of these constants defined elsewhere?

juliapampus replied:
see org.eclipse.edc.protocol.dsp.transform.transformer.PropertyAndTypeNames @janpmeyer

janpmeyer replied:
Constants duplicated Constants used from org.eclipse.edc.protocol.dsp.transform.transformer.PropertyAndTypeNames

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,81 @@
+
+        builder.add(DSPACE_CONTRACTAGREEMENT_TYPE, transferRequestMessage.getContractId());
+        builder.add(DCT_FORMAT, transferRequestMessage.getDataDestination().getType());
+        builder.add(DSPACE_DATAADDRESS_TYPE, transformDataAddress(transferRequestMessage.getDataDestination(), context)); //TODO move to separate transformer
```
Please resolve the TODO

janpmeyer replied:
TODO solved

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,85 @@
+        return builder.build();
+    }
+
+    private @Nullable JsonObject transformDataAddress(DataAddress address, TransformerContext context) {
```
Please resolve this

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,81 @@
+        return builder.build();
+    }
+
+    private @Nullable JsonObject transformDataAddress(DataAddress address, TransformerContext context) {
```
This method design needs to be modified: if there is no data address, the data address field on the method should not be set; otherwise it will contain an empty JsonObject, which is incorrect.

janpmeyer replied:
Solved with checking of numbers of properties bigger than 1

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferStartMessageTransformer.java:
```diff
@@ -0,0 +1,53 @@
+    public @Nullable JsonObject transform(@NotNull TransferStartMessage transferStartMessage, @NotNull TransformerContext context) {
+        var builder = jsonBuilderFactory.createObjectBuilder();
+
+        builder.add(ID, String.valueOf(UUID.randomUUID()));
```
please use .toString() on the UUID class

janpmeyer replied:
solved

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferStartMessageTransformer.java:
```diff
@@ -0,0 +1,53 @@
+        builder.add(TYPE, DSPACE_TRANSFER_START_TYPE);
+        builder.add(DSPACE_PROCESSID_TYPE, transferStartMessage.getProcessId());
+
+        //TODO ADD missing fields dataAddress from spec
```
Please resolve the TODO

janpmeyer replied:
Added issue

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/from/JsonObjectFromTransferTerminationMessageTransformer.java:
```diff
@@ -0,0 +1,54 @@
+        builder.add(TYPE, DSPACE_TRANSFER_TERMINATION_TYPE);
+        builder.add(DSPACE_PROCESSID_TYPE, transferTerminationMessage.getProcessId());
+
+        //TODO Add field when Message is evolved (code, reason)
```
Please resolve the TODO

janpmeyer replied:
Added issue

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferCompletionMessageTransformer.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public @Nullable TransferCompletionMessage transform(@NotNull JsonObject jsonObject, @NotNull TransformerContext context) {
+        var type = nodeType(jsonObject, context);
+
```
spacing

janpmeyer replied:
line deleted

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferCompletionMessageTransformer.java:
```diff
@@ -0,0 +1,57 @@
+
+            return transferCompletionMessageBuilder.build();
+        } else {
+            context.reportProblem(format("Cannot transform type %s to TransferRequestMessage", type));
```
Why could it not transform? Please add information that will be useful in diagnosing problems

janpmeyer replied:
Removed

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,78 @@
+
+            return transferRequestMessageBuilder.build();
+        } else {
+            context.reportProblem(format("Cannot transform type %s to TransferRequestMessage", type));
```
Same comment about adding useful diagnostic information

janpmeyer replied:
Check removed

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferRequestMessageTransformer.java:
```diff
@@ -0,0 +1,78 @@
+            dataAddressJsonObject.entrySet().forEach((k) -> dataAddressBuilder.property(k.getKey(), dataAddressJsonObject.getString(k.getKey())));
+        }
+
+        //TODO Check for missing items
```
Please resolve

janpmeyer replied:
Resolved

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferStartMessageTransformer.java:
```diff
@@ -0,0 +1,58 @@
+            transformString(jsonObject.get(DSPACE_PROCESSID_TYPE), transferStartMessageBuilder::processId, context);
+            transformString(jsonObject.get(DSPACE_CALLBACKADDRESS_TYPE), transferStartMessageBuilder::callbackAddress, context);
+            //TODO ADD missing fields dataAddress from spec
+
```
Please resolve

janpmeyer replied:
Added Issue

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferStartMessageTransformer.java:
```diff
@@ -0,0 +1,58 @@
+
+            return transferStartMessageBuilder.build();
+        } else {
+            context.reportProblem(format("Cannot transform type %s to TransferRequestMessage", type));
```
useful information

janpmeyer replied:
Check removed

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferTerminationMessageTransformer.java:
```diff
@@ -0,0 +1,58 @@
+
+            return transferTerminationMessageBuilder.build();
+        } else {
+            context.reportProblem(format("Cannot transform type %s to TransferRequestMessage", type));
```
useful information?

janpmeyer replied:
Check removed

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferCompletionMessageTransformer.java:
```diff
@@ -0,0 +1,57 @@
+        var type = nodeType(jsonObject, context);
+
+
+        if (DSPACE_TRANSFER_COMPLETION_TYPE.equals(type)) {
```
Many of these transformers have the same implementation pattern: they check the type and register a problem if it doesn't match. Aren't these transformers dispatched to by type? If so, these should be refactored to avoid this and perform a one-line assertion at the start of the method. Currently, these checks are cluttering all of the implementations.

jimmarino replied:
Let's take `JsonObjectToTransferCompletionMessageTransformer` as an example.

A DSP JSON-LD message is deserialized off the wire by Jersey and passed to the DSP controller. The controller needs to validate the incoming message type is correct (or delegate validation). If the message type is incorrect, it should respond with a BAD_REQUEST since the client sent an invalid request according to the DSP specifications. The transformers should not perform this type of error handling as it will result in awkward code and difficulty propagating sufficient error information back to the controller so that it can send the appropriate response to the client. 
 
In other words, the transformers may encounter other problems, but each parent (the controller or a recursively invoked transformer) should ensure valid type dispatching. This will greatly simplify the transformer implementations.

janpmeyer replied:
I see your point, but can we do it in an additional Issue and refactor this later.

Because this implementation pattern is already used in the dsp-catalog-transform modul and dsp-transform modul.

And do it in the same way for all the transform modules in the dsp?

jimmarino replied:
We should fix it now so as not to introduce more technical debt. Those other examples slipped through and they also need to be fixed.

janpmeyer replied:
Removed check, weill be done in API

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferStartMessageTransformer.java:
```diff
@@ -0,0 +1,58 @@
+        var type = nodeType(jsonObject, context);
+
+
+        if (DSPACE_TRANSFER_START_TYPE.equals(type)) {
```
same comments as before

jimmarino commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/test/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/JsonObjectToTransferProcessTransformerTest.java:
```diff
@@ -0,0 +1,204 @@
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_TRANSFER_START_TYPE;
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_TRANSFER_TERMINATION_TYPE;
+
+public class JsonObjectToTransferProcessTransformerTest {
```
refactor into separate unit tests

jimmarino reviewed (changes requested):
Please see my comments inline, particularly:

1. Cleaning up the implementations to not check on type (assuming that is used to dispatch to) and substituting a simple assertion method.
2. Resolving all TODOs in an appropriate way. In general, code should not be merged with TODOs. If something must be deferred to another PR, file an issue and put a comment in the code pointing to the PR.
3. Refactor to unit tests
4. Add appropriate diagnostic information

janpmeyer commented:
Hey. Hope all the wanted changes are gone into the code.

> 1. Cleaning up the implementations to not check on type (assuming that is used to dispatch to) and substituting a simple assertion method.

I hope i changed it the way you meant it.

> 2. Resolving all TODOs in an appropriate way. In general, code should not be merged with TODOs. If something must be deferred to another PR, file an issue and put a comment in the code pointing to the PR.

All TODOs with no related issue should be gone.

> 3. Refactor to unit tests

Is done

> 4. Please justify the addition of an external dependency

The dependencies werent necessary. My bad. I removed them.

> 5. Add appropriate diagnostic information

Because of changing the implementation pattern in 1. this information is gone.

ndr-brt commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/DspTransferProcessTransformExtension.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Map;
+
+@Extension(value = JsonLdExtension.NAME)
+@Provides({JsonLdTransformerRegistry.class})
```
not true, this class does not provide that registry, it requires it (the requirement is already descibed by the `@Inject` annotation

janpmeyer replied:
Annotation removed. Sorry forgot it.

ndr-brt commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/test/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/from/JsonObjectFromDataAddressTransformerTest.java:
```diff
@@ -0,0 +1,71 @@
+
+        var result = transformer.transform(message, context);
+
+        Assertions.assertNotNull(result);
```
please use `assertJ` methods (`assertThat(result).isNotNull()`), same for all the other test classes

janpmeyer replied:
Also changed.

janpmeyer commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/build.gradle.kts:
```diff
@@ -0,0 +1,38 @@
+    implementation("jakarta.json:jakarta.json-api:2.1.1")
+    implementation("org.eclipse.parsson:parsson:1.1.1")
+
+    implementation(libs.dnsOverHttps)
```
Dependency removed

janpmeyer commented on data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/build.gradle.kts:
```diff
@@ -0,0 +1,38 @@
+
+repositories {
+    mavenCentral()
+}
```
Removed

jimmarino reviewed (approved):
I had a few minor nits about spacing and adding documentation to the extension. If you can quiuckly fix those, this should be merged in.
